src/raftstore/store/store.rs

```diff
@@ -1456,6 +1619,255 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+                    .unwrap()
+            };
+
+            let mut request = new_admin_request(sibling_region.get_id(), sibling_peer.peer.clone());
+            request
+                .mut_header()
+                .set_region_epoch(sibling_region.get_region_epoch().clone());
+            let mut admin = AdminRequest::new();
+            admin.set_cmd_type(AdminCmdType::CommitMerge);
+            admin.mut_commit_merge().set_source(region.clone());
+            admin.mut_commit_merge().set_commit(state.get_commit());
+            admin
+                .mut_commit_merge()
+                .set_entries(RepeatedField::from_vec(entries));
+            request.set_admin_request(admin);
+            request
+        };
+        // Please note that, here assumes that the unit of network isolation is store rather than
+        // peer. So a quorum stores of souce region should also be the quorum stores of target
+        // region. Otherwise we need to enable proposal forwarding.
+        self.propose_raft_command(req, Callback::None);
+        Ok(())
+    }
+
+    fn rollback_merge(&mut self, region: &metapb::Region) {
+        let req = {
+            let peer = &self.region_peers[&region.get_id()];
+            let state = peer.pending_merge.as_ref().unwrap();
+            let mut request = new_admin_request(region.get_id(), peer.peer.clone());
+            request
+                .mut_header()
+                .set_region_epoch(peer.region().get_region_epoch().clone());
+            let mut admin = AdminRequest::new();
+            admin.set_cmd_type(AdminCmdType::RollbackMerge);
+            admin.mut_rollback_merge().set_commit(state.get_commit());
+            request.set_admin_request(admin);
+            request
+        };
+        self.propose_raft_command(req, Callback::None);
+    }
+
+    fn on_check_merge(&mut self, event_loop: &mut EventLoop<Self>) {
+        let merging_regions = self.merging_regions.take().unwrap();
+        for region in &merging_regions {
+            if let Err(e) = self.schedule_merge(region) {
+                info!(
+                    "[region {}] failed to schedule merge, rollback: {:?}",
+                    region.get_id(),
+                    e
+                );
+                self.rollback_merge(region);
+            }
+        }
+        self.merging_regions = Some(merging_regions);
+        self.register_merge_check_tick(event_loop);
+    }
+
+    fn on_ready_prepare_merge(&mut self, region: metapb::Region, state: MergeState) {
+        {
+            let peer = self.region_peers.get_mut(&region.get_id()).unwrap();
+            peer.pending_merge = Some(state);
+            peer.mut_store().region = region.clone();
+        }
+
+        if let Err(e) = self.schedule_merge(&region) {
+            info!(
+                "[region {}] failed to schedule merge, rollback: {:?}",
+                region.get_id(),
+                e
+            );
+            self.rollback_merge(&region);
```

Comment:
need return here?

Reply:
No, the region can only be popped after `rollback` is applied.